README.md
# angular-stellar

[![Gem Version](https://badge.fury.io/rb/angular-stellar.png)](http://badge.fury.io/rb/angular-stellar) [![Bower version](https://badge.fury.io/bo/angular-stellar.png)](http://badge.fury.io/bo/angular-stellar) [![NPM version](https://badge.fury.io/js/angular-stellar.png)](http://badge.fury.io/js/angular-stellar) [![Build Status](https://secure.travis-ci.org/tomchentw/angular-stellar.png)](http://travis-ci.org/tomchentw/angular-stellar) [![Coverage Status](https://coveralls.io/repos/tomchentw/angular-stellar/badge.png)](https://coveralls.io/r/tomchentw/angular-stellar) [![Code Climate](https://codeclimate.com/github/tomchentw/angular-stellar.png)](https://codeclimate.com/github/tomchentw/angular-stellar)  [![Dependency Status](https://gemnasium.com/tomchentw/angular-stellar.png)](https://gemnasium.com/tomchentw/angular-stellar)

Rewrite of Stellar.js native AngularJS directives without jQuery dependency


## Project philosophy

### Develop in LiveScript
[LiveScript](http://livescript.net/) is a compile-to-js language, which provides us more robust way to write JavaScript.  
It also has great readibility and lots of syntax sugar just like you're writting python/ruby.

### Native, lightweight directives
`Stellar.js` provides the same interface with angular `directives`.  
We use the similarity between them and provides seamless intergration with `Stellar.js`.  


## Installation

This project follows **DRY** principle and has two dependencies, which are `angularjs` and `ui.bootstrap.position`.

### Just use it

* Download and include [`position.js`](https://github.com/angular-ui/bootstrap/blob/master/src/position/position.js).
* Download and include [`angular-stellar.js`](https://github.com/tomchentw/angular-stellar/blob/master/angular-stellar.js) OR [`angular-stellar.min.js`](https://github.com/tomchentw/angular-stellar/blob/master/angular-stellar.min.js).  

Then include them through script tag in your HTML.

### **Rails** projects (Only support 3.1+)
Add this line to your application's Gemfile:
```ruby
gem 'angular-stellar'
```

And then execute:

    $ bundle

If you are NOT using [`angular-ui/bootstrap`](https://github.com/angular-ui/bootstrap), you have to download
[`position.js`](https://github.com/angular-ui/bootstrap/blob/master/src/position/position.js) and put it in  `vendor/assets/javascripts/`

Then add these lines to the top of your `app/assets/javascripts/application.js` file:

```javascript
//= require angular
//= require ui-bootstrap-tpls
(`//= require position` instead if you aren't using boostrap)
//= require angular-stellar
```

And include in your `angular` module definition:
    
    /* 'angular.stellar' DO depend on 'ui.bootstrap.position' module.
     */    
    var module = angular.module('my-awesome-project', [/* 'bootstrap', */ 'angular.stellar']).


## Usage

### "stellar-ratio": Make the element has parallax effect on front

### "stellar-background-ratio": Make the element has parallax effect on background

### "stellarConfig": constant for configuration

#### scrollProperty
_default_ to: 'scroll'

Available options are 'position', 'margin', 'transform', or you can define it yourself.

#### positionProperty
_default_ to: 'position'

Available options are 'transform', or you can define it yourself.

#### horizontalScrolling
_default_ to: true

Whether or not to enable scrolling in horizontal, only apply to `stellar-ratio` elements.

#### verticalScrolling
_default_ to: true

Whether or not to enable scrolling in vertical, only apply to `stellar-ratio` elements.

#### horizontalOffset
_default_ to: 0

Global offset added to each element.

#### verticalOffset
_default_ to: 0

Global offset added to each element.

#### hideDistantElements
_default_ to: true

When `stellar-ratio` elements are scrolled outside of visible area, hide them.

#### hideElement
_default_ to: `function(it){ it.addClass('ng-hide'); }`

You can have your implementation on your own.

#### showElement
_default_ to: `function(it){ it.removeClass('ng-hide'); }`

You can have your implementation on your own.

#### isElementHidden
_default_ to: `function(it){ return it.hasClass('ng-hide'); }`

You can have your implementation on your own.


## Contributing

[![devDependency Status](https://david-dm.org/tomchentw/angular-stellar/dev-status.png?branch=master)](https://david-dm.org/tomchentw/angular-stellar#info=devDependencies)

1. Fork it ( http://github.com/tomchentw/angular-stellar/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
<a name="0.2.6"></a>
### 0.2.6 (2014-01-17)


#### Bug Fixes

* **Target.handleUpdate:** fix refer to wrong variable bug ([77398653630fbcf09d24059ad80cf55c5f9f5422](git://github.com/tomchentw/angular-stellar.git/commit/77398653630fbcf09d24059ad80cf55c5f9f5422))
* **gemspec:** add railties as runtime dependency ([ca9c5060d03f4448ed4f89326ac98e08fdbb7055](git://github.com/tomchentw/angular-stellar.git/commit/ca9c5060d03f4448ed4f89326ac98e08fdbb7055))
* **gulpfile:** add release-git task with `git push`, `--tags` ([835f72be78644d47713b0a2d221a684500722d61](git://github.com/tomchentw/angular-stellar.git/commit/835f72be78644d47713b0a2d221a684500722d61))
* **travis:** add firefox start script ([5352f2be094e81903a9cbea20c1d402a27841c43](git://github.com/tomchentw/angular-stellar.git/commit/5352f2be094e81903a9cbea20c1d402a27841c43))


#### Features

* **npm:** add .npmignore and add release script to npm ([4f82f576e4a3e4bb076fde619abba2e319a1bc54](git://github.com/tomchentw/angular-stellar.git/commit/4f82f576e4a3e4bb076fde619abba2e319a1bc54))


<a name="0.2.5"></a>
### 0.2.5 (2014-01-16)


#### Bug Fixes

* **gemspec:** add railties as runtime dependency ([ca9c5060d03f4448ed4f89326ac98e08fdbb7055](git://github.com/tomchentw/angular-stellar.git/commit/ca9c5060d03f4448ed4f89326ac98e08fdbb7055))


<a name="0.2.4"></a>
### 0.2.4 (2014-01-16)


#### Bug Fixes

* **travis:** add firefox start script ([5352f2be094e81903a9cbea20c1d402a27841c43](git://github.com/tomchentw/angular-stellar.git/commit/5352f2be094e81903a9cbea20c1d402a27841c43))


<a name="0.2.3"></a>
### 0.2.3 (2014-01-13)


#### Features

* **npm:** add .npmignore and add release script to npm ([4f82f576e4a3e4bb076fde619abba2e319a1bc54](git://github.com/tomchentw/angular-stellar.git/commit/4f82f576e4a3e4bb076fde619abba2e319a1bc54))


<a name="0.2.2"></a>
### 0.2.2 (2014-01-13)


<a name="0.2.1"></a>
### 0.2.1 (2014-01-13)


#### Bug Fixes

* **Target.handleUpdate:** fix refer to wrong variable bug ([77398653630fbcf09d24059ad80cf55c5f9f5422](git://github.com/tomchentw/angular-stellar.git/commit/77398653630fbcf09d24059ad80cf55c5f9f5422))


<a name="0.2.0"></a>
## 0.2.0 (2014-01-13)


#### Features

* **$requestAF:** add implementation ([679cd710d87f40c95fdffb5d485adab472f91bce](git://github.com/tomchentw/angular-stellar.git/commit/679cd710d87f40c95fdffb5d485adab472f91bce))
* **angular-stellar:** done with refactor ([d1fcc05f0ae7f1b00a574e301bb9564f30da485a](git://github.com/tomchentw/angular-stellar.git/commit/d1fcc05f0ae7f1b00a574e301bb9564f30da485a))
* **transform:** add transform support ([aa6b1bf337c628ba9c5288ec02d7287de35d6b0f](git://github.com/tomchentw/angular-stellar.git/commit/aa6b1bf337c628ba9c5288ec02d7287de35d6b0f))



test/karma.conf.js
/*
 * https://github.com/karma-runner/karma/blob/master/test/client/karma.conf.js
 */
module.exports = function(config) {
  config.set({
    // base path, that will be used to resolve files and exclude
    basePath: '../',

    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
      'bower_components/jquery/jquery.js',
      'bower_components/angular/angular.js',
      'bower_components/angular-mocks/angular-mocks.js',
      'bower_components/angular-bootstrap/ui-bootstrap.js',
      'tmp/angular-stellar.js',
      'tmp/angular-stellar.spec.js'
    ],

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress'
    // CLI --reporters progress
    reporters: ['progress', 'coverage'],

    preprocessors: {
      // source files, that you wanna generate coverage for
      // do not include tests or libraries
      // (these files will be instrumented by Istanbul)
      'tmp/angular-stellar.js': ['coverage']
    },

    // web server port
    // CLI --port 9876
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    // CLI --colors --no-colors
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    // CLI --log-level debug
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: false,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    // CLI --browsers Chrome,Firefox,Safari
    browsers: [process.env.TRAVIS ? 'Firefox' : 'Chrome'],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 20000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: true,

    // report which specs are slower than 500ms
    // CLI --report-slower-than 500
    reportSlowerThan: 500,

     // optionally, configure the reporter
    coverageReporter: {
      type : 'lcovonly',
      dir : 'coverage/'
    },

    plugins: [
      'karma-jasmine',
      'karma-chrome-launcher',
      'karma-firefox-launcher',
      'karma-coverage'
    ]
  });
};


test/protractor.conf.js
// A reference configuration file.
var baseConfig = exports.config = {
  // ----- How to setup Selenium -----
  //
  // There are three ways to specify how to use Selenium. Specify one of the
  // following:
  //
  // 1. seleniumServerJar - to start Selenium Standalone locally.
  // 2. seleniumAddress - to connect to a Selenium server which is already
  //    running.
  // 3. sauceUser/sauceKey - to use remote Selenium servers via SauceLabs.
  //
  // If the chromeOnly option is specified, no Selenium server will be started,
  // and chromeDriver will be used directly (from the location specified in
  // chromeDriver)

  // The location of the selenium standalone server .jar file, relative
  // to the location of this config. If no other method of starting selenium
  // is found, this will default to
  // node_modules/protractor/selenium/selenium-server...
  seleniumServerJar: null,
  // The port to start the selenium server on, or null if the server should
  // find its own unused port.
  seleniumPort: null,
  // Chromedriver location is used to help the selenium standalone server
  // find chromedriver. This will be passed to the selenium jar as
  // the system property webdriver.chrome.driver. If null, selenium will
  // attempt to find chromedriver using PATH.
  chromeDriver: null,
  // If true, only chromedriver will be started, not a standalone selenium.
  // Tests for browsers other than chrome will not run.
  chromeOnly: false,
  // Additional command line options to pass to selenium. For example,
  // if you need to change the browser timeout, use
  // seleniumArgs: ['-browserTimeout=60'],
  seleniumArgs: [],

  // If sauceUser and sauceKey are specified, seleniumServerJar will be ignored.
  // The tests will be run remotely using SauceLabs.
  sauceUser: null,
  sauceKey: null,

  // The address of a running selenium server. If specified, Protractor will
  // connect to an already running instance of selenium. This usually looks like
  // seleniumAddress: 'http://localhost:4444/wd/hub'
  seleniumAddress: null,

  // The timeout for each script run on the browser. This should be longer
  // than the maximum time your application needs to stabilize between tasks.
  allScriptsTimeout: 11000,

  // ----- What tests to run -----
  //
  // Spec patterns are relative to the location of this config.
  specs: [
    '../tmp/angular-stellar.scenario.js'
  ],

  // ----- Capabilities to be passed to the webdriver instance ----
  //
  // For a full list of available capabilities, see
  // https://code.google.com/p/selenium/wiki/DesiredCapabilities
  // and
  // https://code.google.com/p/selenium/source/browse/javascript/webdriver/capabilities.js
  capabilities: {
    'browserName': 'chrome'
  },

  // ----- More information for your tests ----
  //
  // A base URL for your application under test. Calls to protractor.get()
  // with relative paths will be prepended with this.
  baseUrl: 'http://localhost:2999',

  // Selector for the element housing the angular app - this defaults to
  // body, but is necessary if ng-app is on a descendant of <body>  
  rootElement: 'body',

  // A callback function called once protractor is ready and available, and
  // before the specs are executed
  // You can specify a file containing code to run by setting onPrepare to
  // the filename string.
  onPrepare: function() {
    // At this point, global 'protractor' object will be set up, and jasmine
    // will be available. For example, you can add a Jasmine reporter with:
    //     jasmine.getEnv().addReporter(new jasmine.JUnitXmlReporter(
    //         'outputdir/', true, true));
  },

  // The params object will be passed directly to the protractor instance,
  // and can be accessed from your test. It is an arbitrary object and can
  // contain anything you may need in your test.
  // This can be changed via the command line as:
  //   --params.login.user 'Joe'
  params: {
    login: {
      user: 'Jane',
      password: '1234'
    }
  },
  
  // ----- Options to be passed to minijasminenode -----
  //
  // See the full list at https://github.com/juliemr/minijasminenode
  jasmineNodeOpts: {
    // onComplete will be called just before the driver quits.
    onComplete: null,
    // If true, display spec names.
    isVerbose: false,
    // If true, print colors to the terminal.
    showColors: true,
    // If true, include stack traces in failures.
    includeStackTrace: true,
    // Default time to wait in ms before a test fails.
    defaultTimeoutInterval: 30000
  }
};

if (process.env.TRAVIS) {
  baseConfig.sauceUser = process.env.SAUCE_USERNAME;
  baseConfig.sauceKey = process.env.SAUCE_ACCESS_KEY;

  baseConfig.capabilities['tunnel-identifier'] = process.env.TRAVIS_JOB_NUMBER;
} else {// local
  baseConfig.seleniumServerJar = '/usr/local/lib/node_modules/protractor/selenium/selenium-server-standalone-2.39.0.jar';
  baseConfig.chromeDriver = '/usr/local/lib/node_modules/protractor/selenium/chromedriver';
}






Gruntfile.js
/*global module:false*/
module.exports = function(grunt) {
  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-karma');
  
  grunt.loadNpmTasks('grunt-livescript');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-copy');
  // Project configuration.
  /*jshint scripturl:true*/
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON('package.json'),
    fdr: {
      src:    'src/',
      dest:   './',
      demo: 'demo/',
      tmp: 'tmp/'
    },
    banner: '' +
      '/*! <%= pkg.title || pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? " * " + pkg.homepage + "\\n" : "" %>' +
      ' * Copyright (c) <%= grunt.template.today("yyyy") %> [<%= pkg.author.name %>](<%= pkg.author.url %>);\n' +
      ' * Licensed [<%= pkg.license.type %>](<%= pkg.license.url %>)\n' +
      ' */\n',
    // Task configuration.
    delta: {
      ls: {
        files: ['<%= fdr.src %><%= pkg.name %>.*.ls'],
        tasks: ['livescript:watch']
      },
      js: {
        files: ['<%= fdr.tmp %><%= pkg.name %>.*.ls'],
        tasks: ['karma:watch:run']
      }
    },
    livescript: { watch: {
        expand: true,
        cwd: '<%= fdr.src %>',
        src: '<%= pkg.name %>.*.ls',
        dest: '<%= fdr.tmp %>',
        filter: 'isFile',
        options: {
          bare: true
        }
      },          compile: {
        src: '<%= fdr.src %><%= pkg.name %>.js.ls',
        dest: '<%= fdr.dest %><%= pkg.name %>.js'
      }
    },
    uglify: { compile: {
        src: '<%= livescript.compile.dest %>',
        dest: '<%= grunt.config.get("livescript.compile.dest").replace(".js", ".min.js") %>'
      },
      options: { banner: '<%= banner %>' }
    },
    copy: { rubygem: {
        expand: true,
        cwd: '<%= fdr.dest %>',
        src: '<%= pkg.name %>.js',
        dest: 'vendor/assets/javascripts/',
        filter: 'isFile'
      },    demo: {
        expand: true,
        cwd: '<%= fdr.dest %>',
        src: '<%= pkg.name %>.js',
        dest: '<%= fdr.demo %>',
        filter: 'isFile'
      }
    },
    karma: {
      options: {
        frameworks: ['jasmine'],
        files: [
          'misc/test-lib/jquery-1.10.2.min.js',
          'misc/test-lib/angular.1.2.6.min.js',
          'misc/test-lib/angular-mocks.1.2.6.min.js',
          'misc/test-lib/position.js',
          '<%= fdr.tmp %><%= pkg.name %>.js.ls',
          '<%= fdr.tmp %><%= pkg.name %>.spec.ls'
        ],
        browsers: ['Chrome'],
        port: 9018,
        runnerPort: 9100,
        colors: true,
        autoWatch: false,
        singleRun: false
      },
      watch: {
        background: true
      },
      travis: {
        singleRun: true,
        browsers: ['Firefox']
      },
      continuous: {
        singleRun: true
      }
    },
  });
  //
  grunt.renameTask('watch', 'delta');
  grunt.registerTask('watch', ['livescript:watch', 'karma:watch', 'delta']);
  //
  grunt.registerTask('default', ['livescript:compile', 'uglify:compile', 'copy'])
};


angular-stellar.js
/*! tomchentw-boilerplate - v 0.2.5 - Thu Jan 16 2014 01:17:47 GMT+0800 (CST)
 * https://github.com/tomchentw/angular-stellar
 * Copyright (c) 2014 [tomchentw](https://github.com/tomchentw/);
 * Licensed [MIT](http://tomchentw.mit-license.org/)
 *//*global angular:false*/
(function(){
  var bind, noop, extend, delayInFPS, $requestAnimationFrame, stellarConfig, Target, $css, $vendorPrefix, stellarTarget, stellarBackgroundRatio, stellarRatio;
  bind = angular.bind, noop = angular.noop, extend = angular.extend;
  delayInFPS = 1000 / 60;
  $requestAnimationFrame = ['$window', '$log'].concat(function($window, $log){
    var NAMES, i$, len$, name, that;
    NAMES = ['request', 'webkitRequest', 'mozRequest', 'oRequest', 'msRequest'];
    for (i$ = 0, len$ = NAMES.length; i$ < len$; ++i$) {
      name = NAMES[i$];
      if (that = $window[name + "AnimationFrame"]) {
        return bind($window, that);
      }
    }
    $log.warn('Can\'t find requestAnimationFrame in your browser. Use polyfill.');
    return function(it){
      return $window.setTimeout(it, delayInFPS);
    };
  });
  stellarConfig = {
    scrollProperty: 'scroll',
    positionProperty: 'position',
    horizontalScrolling: true,
    verticalScrolling: true,
    horizontalOffset: 0,
    verticalOffset: 0,
    hideDistantElements: true,
    hideElement: function(it){
      it.addClass('ng-hide');
    },
    showElement: function(it){
      it.removeClass('ng-hide');
    },
    isElementHidden: function(it){
      return it.hasClass('ng-hide');
    }
  };
  Target = (function(){
    Target.displayName = 'Target';
    var prototype = Target.prototype, constructor = Target;
    Target._lastTime = 0;
    Target._targets = [];
    Target.getInstance = function(name, $element){
      return Target[name] || new Target(name, $element);
    };
    Target.handleUpdate = function(timestamp){
      var justUpdated, rescheduled, i$, ref$, len$, target;
      justUpdated = timestamp - this._lastTime < delayInFPS;
      if (justUpdated) {
        return true;
      }
      this._lastTime = timestamp;
      rescheduled = false;
      for (i$ = 0, len$ = (ref$ = this._targets).length; i$ < len$; ++i$) {
        target = ref$[i$];
        if (target.handleUpdate() && !rescheduled) {
          rescheduled = true;
        }
      }
      return rescheduled;
    };
    function Target(name, $element){
      constructor[name] = this;
      constructor._targets.push(this);
      this._$element = $element;
      this._callbacks = [];
      this._props = {};
    }
    prototype.addCallbak = function(it){
      var index;
      index = -1 + this._callbacks.push(it);
      return bind(this._callbacks, this._callbacks.splice, index);
    };
    prototype.handleUpdate = function(){
      var updated, i$, ref$, len$, callback;
      updated = this.getOffset();
      updated.scrollTop = this.getTop();
      updated.scrollLeft = this.getLeft();
      if (angular.equals(updated, this._props)) {
        return;
      }
      extend(this._props, updated);
      for (i$ = 0, len$ = (ref$ = this._callbacks).length; i$ < len$; ++i$) {
        callback = ref$[i$];
        callback(this._props);
      }
    };
    prototype.handleWebkitBug = function(){
      var ref$, scrollTop, scrollLeft;
      ref$ = this._props, scrollTop = ref$.scrollTop, scrollLeft = ref$.scrollLeft;
      this.setTop(scrollTop + 1);
      this.setLeft(scrollLeft + 1);
      this.setTop(scrollTop);
      this.setLeft(scrollLeft);
    };
    prototype.getLeft = function(){
      return this._$element.prop('scrollLeft');
    };
    prototype.setLeft = function(it){
      this._$element.prop('scrollLeft', it);
    };
    prototype.getTop = function(){
      return this._$element.prop('scrollTop');
    };
    prototype.setTop = function(it){
      this._$element.prop('scrollTop', it);
    };
    return Target;
  }());
  $css = ['$window'].concat(function($window){
    return function($element, cssprop){
      var el, style;
      el = $element[0];
      style = el.currentStyle
        ? el.currentStyle
        : $window.getComputedStyle
          ? $window.getComputedStyle(el)
          : el.style;
      if (cssprop) {
        return style[cssprop];
      } else {
        return style;
      }
    };
  });
  $vendorPrefix = ['$css'].concat(function($css){
    var VENDORS, name, that, prefix;
    VENDORS = /^(moz|webkit|khtml|o|ms)(?=[A-Z])/;
    for (name in $css(angular.element('<script></script>'))) {
      if (that = name.match(VENDORS)) {
        prefix = that[0];
        break;
      }
    }
    return function(cssprop){
      return prefix + (!prefix
        ? cssprop
        : angular.uppercase(cssprop.charAt(0)) + "" + cssprop.slice(1));
    };
  });
  stellarTarget = ['$window', '$document', '$position', '$requestAnimationFrame', '$css', '$vendorPrefix', 'stellarConfig'].concat(function($window, $document, $position, $requestAnimationFrame, $css, $vendorPrefix, stellarConfig){
    var windowTarget, documentTarget, scrollProperty, schedule;
    windowTarget = Target.getInstance('window', angular.element($window));
    documentTarget = Target.getInstance('document', $document);
    windowTarget.getOffset = documentTarget.getOffset = function(){
      var docEl;
      docEl = $document[0].documentElement;
      return {
        height: docEl.clientHeight,
        width: docEl.clientWidth,
        top: 0,
        left: 0
      };
    };
    windowTarget.getLeft = documentTarget.getLeft = function(){
      return $window.pageXOffset;
    };
    windowTarget.getTop = documentTarget.getTop = function(){
      return $window.pageYOffset;
    };
    windowTarget.setLeft = documentTarget.setLeft = function(it){
      $window.scrollTo(it, $window.pageYOffset);
    };
    windowTarget.setTop = documentTarget.setLeft = function(it){
      $window.scrollTo($window.pageXOffset, it);
    };
    scrollProperty = stellarConfig.scrollProperty;
    extend(Target.prototype, {
      getOffset: function(){
        return $position.offset(this._$element);
      }
    }, angular.isObject(scrollProperty)
      ? scrollProperty
      : (function(){
        var cssInt, prefixedTransform;
        cssInt = function($element, cssprop){
          return parseInt($css($element, cssprop), 10);
        };
        prefixedTransform = $vendorPrefix('transform');
        switch (scrollProperty) {
        case 'position':
          return {
            getLeft: function(){
              return -1 * cssInt(this._$element, 'left');
            },
            getTop: function(){
              return -1 * cssInt(this._$element, 'top');
            }
          };
        case 'margin':
          return {
            getLeft: function(){
              return -1 * cssInt(this._$element, 'marginLeft');
            },
            getTop: function(){
              return -1 * cssInt(this._$element, 'marginTop');
            }
          };
        case 'transform':
          return {
            getLeft: function(){
              var transform;
              transform = $css(this._$element, prefixedTransform);
              if ('none' === transform) {
                return 0;
              } else {
                return -1 * parseInt(transform.match(/(-?[0-9]+)/g)[4], 10);
              }
            },
            getTop: function(){
              var transform;
              transform = $css(this._$element, prefixedTransform);
              if ('none' === transform) {
                return 0;
              } else {
                return -1 * parseInt(transform.match(/(-?[0-9]+)/g)[5], 10);
              }
            }
          };
        case 'scroll':
          // fallthrough
        default:
          return {};
        }
      }.call(this)));
    schedule = bind(this, $requestAnimationFrame, function(timestamp){
      var reschedule;
      reschedule = Target.handleUpdate(timestamp);
      if ('scroll' === scrollProperty && !reschedule) {
        return;
      }
      schedule();
    });
    if (scrollProperty === 'scroll') {
      windowTarget._$element.on('scroll resize', schedule);
    }
    return function(name, $element){
      return schedule() && Target.getInstance(name, $element);
    };
  });
  stellarBackgroundRatio = ['$position', '$css', 'stellarTarget'].concat(function($position, $css, stellarTarget){
    var getBackgroundPosition, computeRatio;
    getBackgroundPosition = function($element){
      var bgPos;
      bgPos = $css($element, 'backgroundPosition').split(' ');
      return [parseInt(bgPos[0], 10), parseInt(bgPos[1], 10)];
    };
    computeRatio = function($element, $attrs){
      var stellarBackgroundRatio, fixedRatioOffset;
      stellarBackgroundRatio = $attrs.stellarBackgroundRatio || 1;
      fixedRatioOffset = 'fixed' === $css($element, 'backgroundAttachment') ? 0 : 1;
      return fixedRatioOffset - stellarBackgroundRatio;
    };
    return {
      restrict: 'A',
      link: function($scope, $element, $attrs){
        var finalRatio, verticalOffset, horizontalOffset, selfProperties, selfBgPositions, parentProperties;
        finalRatio = computeRatio($element, $attrs);
        verticalOffset = $attrs.stellarVerticalOffset || stellarConfig.verticalOffset;
        horizontalOffset = $attrs.stellarHorizontalOffset || stellarConfig.horizontalOffset;
        selfProperties = $position.offset($element);
        selfBgPositions = getBackgroundPosition($element);
        parentProperties = {
          top: 0,
          left: 0
        };
        selfProperties.bgTop = selfBgPositions[1];
        selfProperties.bgLeft = selfBgPositions[0];
        $scope.$on('$destroy', stellarTarget('window').addCallbak(function(targetProps){
          var bgTop, bgLeft;
          bgTop = finalRatio * (targetProps.scrollTop + verticalOffset - targetProps.top - selfProperties.top + parentProperties.top - selfProperties.bgTop);
          bgLeft = finalRatio * (targetProps.scrollLeft + horizontalOffset - targetProps.left - selfProperties.left + parentProperties.left - selfProperties.bgLeft);
          $element.css('background-position', bgLeft + "px " + bgTop + "px");
        }));
      }
    };
  });
  stellarRatio = ['$css', '$vendorPrefix', '$position', 'stellarConfig', 'stellarTarget'].concat(function($css, $vendorPrefix, $position, stellarConfig, stellarTarget){
    var positionProperty, horizontalScrolling, verticalScrolling, setPosition, computeIsFixed, computeRatio;
    positionProperty = stellarConfig.positionProperty, horizontalScrolling = stellarConfig.horizontalScrolling, verticalScrolling = stellarConfig.verticalScrolling;
    setPosition = angular.isFunction(positionProperty.setPosition)
      ? positionProperty.setPosition
      : (function(){
        var prefixedTransform, setTop, setLeft;
        prefixedTransform = $vendorPrefix('transform');
        switch (positionProperty) {
        case 'position':
          setTop = function($element, top){
            $element.css('top', top);
          };
          setLeft = function($element, left){
            $element.css('left', left);
          };
          break;
        case 'transform':
          return function($element, left, startingLeft, top, startingTop){
            $element.css(prefixedTransform, "translate3d(" + (left - startingLeft) + "px, " + (top - startingTop) + "px, 0)");
          };
        }
        return function($element, left, startingLeft, top, startingTop){
          if (horizontalScrolling) {
            setLeft($element, left + "px", startingLeft + "px");
          }
          if (verticalScrolling) {
            setTop($element, top + "px", startingTop + "px");
          }
        };
      }.call(this));
    computeIsFixed = function($element){
      return 'fixed' === $css($element, 'position');
    };
    computeRatio = function($element, $attrs){
      var stellarRatio, fixedRatioOffset;
      stellarRatio = $attrs.stellarRatio || 1;
      fixedRatioOffset = computeIsFixed($element) ? 1 : 0;
      return stellarRatio + fixedRatioOffset - 1;
    };
    return {
      restrict: 'A',
      link: function($scope, $element, $attrs){
        var isFixed, finalRatio, verticalOffset, horizontalOffset, selfProperties, selfPositions, parentProperties;
        isFixed = computeIsFixed($element);
        finalRatio = -1 * computeRatio($element, $attrs);
        verticalOffset = $attrs.stellarVerticalOffset || stellarConfig.verticalOffset;
        horizontalOffset = $attrs.stellarHorizontalOffset || stellarConfig.horizontalOffset;
        selfProperties = $position.offset($element);
        selfPositions = $position.position($element);
        parentProperties = {
          top: 0,
          left: 0
        };
        selfProperties.positionTop = selfPositions.top;
        selfProperties.positionLeft = selfPositions.left;
        $scope.$on('$destroy', stellarTarget('window').addCallbak(function(targetProps){
          var newTop, newOffsetTop, newLeft, newOffsetLeft, targetScrollLeft, targetScrollTop, isVisibleHorizontal, isVisibleVertical, isHidden;
          newTop = selfProperties.positionTop;
          newOffsetTop = selfProperties.top;
          if (verticalScrolling) {
            newTop += finalRatio * (targetProps.scrollTop + verticalOffset + targetProps.top + selfProperties.positionTop - selfProperties.top + parentProperties.top);
            newOffsetTop += newTop - selfProperties.positionTop;
          }
          newLeft = selfProperties.positionLeft;
          newOffsetLeft = selfProperties.left;
          if (horizontalScrolling) {
            newLeft += finalRatio * (targetProps.scrollLeft + horizontalOffset + targetProps.left + selfProperties.positionLeft - selfProperties.left + parentProperties.left);
            newOffsetLeft += newLeft - selfProperties.positionLeft;
          }
          if (stellarConfig.hideDistantElements) {
            targetScrollLeft = isFixed
              ? 0
              : targetProps.scrollLeft;
            targetScrollTop = isFixed
              ? 0
              : targetProps.scrollTop;
            isVisibleHorizontal = !horizontalScrolling || (newOffsetLeft + selfProperties.width > targetScrollLeft && newOffsetLeft < targetProps.width + targetProps.left + targetScrollLeft);
            isVisibleVertical = !verticalScrolling || (newOffsetTop + selfProperties.height > targetScrollTop && newOffsetTop < targetProps.height + targetProps.top + targetScrollTop);
          }
          isHidden = stellarConfig.isElementHidden($element);
          if (isVisibleHorizontal && isVisibleVertical) {
            if (isHidden) {
              stellarConfig.showElement($element);
            }
            setPosition($element, newLeft, selfProperties.positionLeft, newTop, selfProperties.positionTop);
          } else if (!isHidden) {
            stellarConfig.hideElement($element);
          }
        }));
      }
    };
  });
  angular.module('angular.stellar', ['ui.bootstrap.position']).constant('stellarConfig', stellarConfig).factory('$requestAnimationFrame', $requestAnimationFrame).factory('$css', $css).factory('$vendorPrefix', $vendorPrefix).factory('stellarTarget', stellarTarget).directive('stellarRatio', stellarRatio).directive('stellarBackgroundRatio', stellarBackgroundRatio);
}).call(this);


misc/karma.conf.js
/*
 * https://github.com/karma-runner/karma/blob/master/test/client/karma.conf.js
 */
module.exports = function(config) {
  config.set({
    // base path, that will be used to resolve files and exclude
    basePath: '../',

    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
      'misc/test-lib/jquery-1.10.2.min.js',
      'misc/test-lib/angular.1.2.6.min.js',
      'misc/test-lib/angular-mocks.1.2.6.min.js',
      'misc/test-lib/position.min.js',
      'tmp/angular-stellar.js',
      'tmp/angular-stellar.spec.js'
    ],

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress'
    // CLI --reporters progress
    reporters: ['progress'],

    // web server port
    // CLI --port 9876
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    // CLI --colors --no-colors
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    // CLI --log-level debug
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: false,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    // CLI --browsers Chrome,Firefox,Safari
    browsers: [process.env.TRAVIS ? 'Firefox' : 'Chrome'],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 20000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: true,

    // report which specs are slower than 500ms
    // CLI --report-slower-than 500
    reportSlowerThan: 500,

    plugins: [
      'karma-jasmine',
      'karma-chrome-launcher',
      'karma-firefox-launcher'
    ]
  });
};
